xdsl/dialects/hw.py

```diff
@@ -112,6 +120,173 @@ def print_parameters(self, printer: Printer) -> None:
         printer.print_string(">")
 
 
+class InnerSymbolTable(OpTrait):
+    """A trait for inner symbol table functionality on an operation.
+    Merges the upstream table of inner symbols and their resolutions and the op trait.
+    """
+
+    def __init__(
+        self, table: Mapping[StringAttr, InnerSymTarget] | None = None
+    ) -> None:
+        if table is None:
+            table = dict()
+        self.symbol_table = table
+        super().__init__()
+
+    @classmethod
+    def verify_region_trait(cls, op: Operation):
+        # Insist that ops with InnerSymbolTable's provide a Symbol, this is
+        # essential to how InnerRef's work.
+        if not op.has_trait(SymbolOpInterface):
+            raise VerifyException(
+                f"Operation {op.name} must have trait SymbolOpInterface"
+            )
```

Comment:
```suggestion
        if not op.has_trait(trait := SymbolOpInterface):
            raise VerifyException(
                f"Operation {op.name} must have trait {trait.__name__}"
            )
```

Can you please do this where appropriate throughout this PR?